### PR TITLE
Pivot record_barcodes table to use one barcode column

### PR DESCRIPTION
**Pull request by jccraft-uw**
In this commit, individual REDCap barcode columns are replaced by a
single barcode column and a redcap_field column to capture the REDCap
field that held the barcode value. This "pivots" the record_barcodes
table from wide to tall, with one barcode per row. There are 3 key
advantages to this approach:
1. The SQL to find a record containing a barcode is greatly simplified.
There's no longer a big `coalesce` that looks through many columns;
instead it looks at only one column.
2. When new REDCap barcode fields need to be added, the SQL won't
have to be modified. The new fields will just need to be incorporated into
`export-record-barcodes`.
3. The user experience is improved for the scenario when a user
dials a barcode and that returns more than one result. Instead of having
to scroll far to the right to see (and try to keep in mind) all the
columns, the user will see the fields that contain the barcode in
real estate that doesn't require scrolling.

Because this pivot takes away the view that shows all the barcodes in all
the columns, the `related` page has been added to show all the barcodes
that belong to the REDCap record(s) to which the input barcode belongs.
This page offers 2 key advantages over the current design:
1. It shows all the barcodes related to the REDCap records, not just all
the barcodes for the particular REDCap instance that the input barcode
belongs to.
2. The page is always available. A Switchboard user can always plug in a
barcode and see its related barcodes. Currently, the similar view is only
available if dialing a barcode returns more than one result.

This commit adds a `unique_record_id` column to the record_barcodes table.
This gives a single column to uniquely identify the REDCap record that a
barcode came from. Having this as a single column simplifies some SQL a bit.
Currently, this is populated with the REDCap project ID |
REDCap record ID. In the event that the group uses multiple REDCap
instances in the future, the REDCap project API URL should be added
to the unique identifier.

The `derived-tables.sql` SQL has been updated to show occurences where the
record_barcodes table contains more than one row for the same REDCap
barcode field for a particular REDCap record instance. This would
indicate a fault in the data refresh. For example, the existing table
might not have gotten dropped before fresh data was added.

### CODE REVIEW

**Review by tsibley — CHANGES_REQUESTED** [review]
Overall looks great! This will be a nice improvement.

It would be nice not to completely duplicate (with a few lines missing) the `templates/page/dial.html` file as `related.html`. Pulling the JS to a shared file and re-using template blocks, for example. But stepping back, do we need the fast, repeat-lookup optimized features of the dialer for this new page? Could we instead do away with the custom `related.html` page and point the home page link at the standard canned-query page instead (`…/sfs-redcap/lookup-barcode?_hide_sql=1`)?

**Comment by jccraft-uw** [comment]
> Overall looks great! This will be a nice improvement.
> 
> It would be nice not to completely duplicate (with a few lines missing) the `templates/page/dial.html` file as `related.html`. Pulling the JS to a shared file and re-using template blocks, for example. But stepping back, do we need the fast, repeat-lookup optimized features of the dialer for this new page? Could we instead do away with the custom `related.html` page and point the home page link at the standard canned-query page instead (`…/sfs-redcap/lookup-barcode?_hide_sql=1`)?

We don't need a custom page at all to show related barcodes. The page doesn't need to do anything fancy. I've removed the `related.html` page and configured the link to use the canned-query page. That's really handy! I also restored the `dial.html` to how it is in the master branch. The only reason I tweaked it was to make it stand out a bit from the `related.html` page--by setting a more descriptive header (h1) and page title, for example. That's no longer needed.

File: bin/export-record-barcodes [hunk 1]
```diff
@@ -1,16 +1,16 @@
...
-from id3c.cli import redcap
-from id3c.json import as_json
-from os import environ
+from sys import stdout, stderr
 from typing import Optional, Dict
 from urllib.parse import urlencode, urljoin
 
+from id3c.cli import redcap
+from id3c.json import as_json
+
```
[the remark this answers is not in the document]
Comment (jccraft-uw) [reply bin/export-record-barcodes]: Right. That was just cleanup. I've moved the cleanup into its own commit. I removed the `import csv` that was an artifact I carried over from a previous incarnation by mistake.

**Comment by tsibley** [comment]
Awesome, thanks!

**Review by tsibley — APPROVED** [review]

**Comment by jccraft-uw** [comment]
Deployed.

**Pull request merged**